File: src/Sequence/world/blocks/imagine/ImagineNode.java
package Sequence.world.blocks.imagine;

import Sequence.world.meta.imagine.BuildingIEc;
import Sequence.world.util.struct.Pair;
import Sequence.world.blocks.imagine.ImagineCenter.ImagineCenterBuild;
import arc.math.Mathf;
import arc.math.geom.Geometry;
import arc.struct.Queue;
import arc.struct.Seq;
import arc.util.Time;
import arc.util.io.Reads;
import mindustry.game.Team;
import mindustry.gen.Building;
import mindustry.graphics.Drawf;
import mindustry.graphics.Pal;
import mindustry.world.Tile;

import static mindustry.Vars.tilesize;
import static mindustry.Vars.world;

public class ImagineNode extends ImagineBlock {
    public int linkRange = 3;

    public ImagineNode(String name) {
        super(name);
    }

    Pair<Building, Integer> findNode(int x, int y, int direction) {
        int maxLen = linkRange + size / 2;
        Building dest = null;
        int dy = Geometry.d4y(direction);
        int dx = Geometry.d4x(direction);
        int offset = size / 2;
        for (int j = 1 + offset; j <= linkRange + offset; ++j) {
            Building other = world.build(x + j * dx, y + j * dy);
            if (other == null) continue;
            if (other instanceof ImagineNodeBuild || other instanceof ImagineCenter.ImagineCenterBuild) {
                maxLen = j;
                dest = other;
                break;
            }
        }
        return new Pair<>(dest, maxLen);
    }

    @Override
    public boolean canPlaceOn(Tile tile, Team team, int rotation) {
        if (tile.x == 60 && tile.y == 60) return true;
        for (int i = 0; i < 4; ++i) {
            if (findNode(tile.x, tile.y, i).getKey() != null) return true;
        }
        return false;
    }

    @Override
    public void drawPlace(int x, int y, int rotation, boolean valid) {
        for (int i = 0; i < 4; ++i) {
            var res = findNode(x, y, i);
            int dx = Geometry.d4x(i), dy = Geometry.d4y(i), maxLen = res.getValue();
            Building dest = res.getKey();
            Drawf.dashLine(Pal.placing,
                    x * tilesize + dx * (tilesize * size / 2f + 2),
                    y * tilesize + dy * (tilesize * size / 2f + 2),
                    x * tilesize + dx * maxLen * tilesize,
                    y * tilesize + dy * maxLen * tilesize
            );
            if (dest != null) {
                Drawf.square(dest.x, dest.y, dest.block.size * tilesize / 2f + 2.5f, 0f);
            }
        }
    }

    public class ImagineNodeBuild extends Building implements BuildingIEc {
        public ImagineCenterBuild center = null;
        protected boolean found = false;

        @Override
        public float capacity() {
            return 0;
        }

        @Override
        public void onRemoved() {
            if (center != null) {
                center.linkedIE.remove(this);
                Seq<Building> all = new Seq<>();
                Queue<Building> queue = new Queue<>();
                queue.add(center);
                while (!queue.isEmpty()) {
                    var e = queue.first();
                    queue.removeLast();
                    for (int i = 0; i < 4; ++i) {
                        Building build = findNode(e.tileX(), e.tileY(), i).getKey();
                        if (build == this) continue;
                        if (build == null) continue;
                        if (all.contains(build)) continue;
                        all.add(build);
                        queue.add(build);
                    }
                }
                for (Building building : center.linkedIE) {
                    if (!all.contains(building)) building.kill();
                }
            }
        }

        @Override
        public void updateTile() {
            if (center == null) {
                if (found) kill();
                else {
                    found = true;
                    findCenter();
                }
            }
            super.updateTile();
        }

        @Override
        public void drawSelect() {
            super.drawSelect();
            float sin = Mathf.sin(Time.time / 15f);
            Drawf.arrow(x, y, center.x, center.y, size * tilesize + sin, 4f + sin);
            Drawf.square(center.x, center.y, center.block.size * tilesize / 2f + 2.5f, 0f);
        }

        private void findCenter() {
            ImagineCenterBuild ce = null;
            for (int i = 0; i < 4; ++i) {
                Building build = findNode(tileX(), tileY(), i).getKey();
                if (build instanceof ImagineCenterBuild c) {
                    if (ce != null && ce != c) kill();
                    ce = c;
                } else if (build instanceof ImagineNodeBuild n) {
                    if (ce != null && ce != n.center) kill();
                    ce = n.center;
                }
            }
            if (ce == null) kill();
            center = ce;
            if (center != null) {
                center.linkedIE.add(this);
            }
        }
    }
}

File: src/sequence/world/blocks/imagine/ImagineNode.kt
package sequence.world.blocks.imagine

import arc.func.Prov
import arc.math.geom.Geometry
import arc.math.geom.Point2
import arc.struct.Seq
import mindustry.Vars
import mindustry.Vars.tilesize
import mindustry.Vars.world
import mindustry.entities.TargetPriority
import mindustry.gen.Building
import mindustry.graphics.Drawf
import mindustry.graphics.Pal
import mindustry.input.Placement
import mindustry.world.blocks.power.PowerNode
import mindustry.world.meta.Env
import sequence.graphic.dashRectBuild
import sequence.util.component1
import sequence.util.component2
import sequence.world.blocks.imagine.ImagineCenter.ImagineCenterBuild
import sequence.world.meta.imagine.ImagineLink
import sequence.world.meta.imagine.toImagineLink
import kotlin.math.abs

class ImagineNode(name: String) : ImagineBlock(name) {
    var range: Int = 5

    init {
        buildType = Prov { ImagineNodeBuild() }
        envEnabled = envEnabled or Env.space
        allowDiagonal = false
        underBullets = true
        priority = TargetPriority.transport
    }

    override fun drawPlace(x: Int, y: Int, rotation: Int, valid: Boolean) {
        for (i in 0..<4) {
            var maxLen = range + size / 2
            var dest: Building? = null
            val dir = Geometry.d4[i]
            val dx = dir.x
            val dy = dir.y
            val offset = size / 2
            for (j in 1 + offset..range + offset) {
                val other = world.build(x + j * dir.x, y + j * dir.y)

                if (other != null && other.isInsulated) {
                    break
                }
                if (other != null && other.block.hasPower && other.team === Vars.player.team() && other.block !is PowerNode) {
                    maxLen = j
                    dest = other
                    break
                }
            }
            Drawf.dashLine(
                Pal.placing,
                x * tilesize + dx * (tilesize * size / 2f + 2),
                y * tilesize + dy * (tilesize * size / 2f + 2),
                (x * tilesize + dx * maxLen * tilesize).toFloat(),
                (y * tilesize + dy * maxLen * tilesize).toFloat()
            )
            if (dest != null) {
                dashRectBuild(Pal.accent, dest)
            }
        }
    }

    override fun changePlacementPath(points: Seq<Point2?>?, rotation: Int, diagonal: Boolean) {
        if (!diagonal) {
            Placement.calculateNodes(
                points, this, rotation
            ) { point: Point2, other: Point2 ->
                abs(point.x - other.x).coerceAtLeast(abs(point.y - other.y)) <= range
            }
        }
    }

    inner class ImagineNodeBuild : Building(), ImagineLink {
        override var center: ImagineCenterBuild? = null

        override fun placed() {
            super.placed()
            checkLink()
        }

        private fun checkLink() {
            for (i in 1..range) {
                for (j in 0..4) {
                    val (dx, dy) = Geometry.d4(j)
                    val tx = tile.x.toInt() + dx * i + (size - 1) / 2
                    val ty = tile.y.toInt() + dy * i + (size - 1) / 2
                    val build = world.build(tx, ty) ?: continue
                    if (build is ImagineCenterBuild) {
                        if (center == null)
                            center = build
                    } else if (build is ImagineLink) {
                        if (center == null)
                            center = build.center
                        else
                            center!!.tryAdd(build.toImagineLink())
                    }
                }
            }
        }

        override fun updateTile() {
            super.updateTile()
            if (center?.tile?.build !== center) center = null
            if (center == null) checkLink()
            center?.links?.add(this.toImagineLink())
        }

        override fun onRemoved() {
            super.onRemoved()
            center?.links?.remove(this.toImagineLink())
        }

        override fun drawSelect() {
            super.drawSelect()
            dashRectBuild(Pal.accent, center ?: return)
        }
    }
}